src/Admin/SiteHealth.php

```diff
@@ -76,11 +76,16 @@ public function __construct( MonitorCssTransientCaching $css_transient_caching,
 	 * Adds the filters.
 	 */
 	public function register() {
-		add_filter( 'site_status_tests', [ $this, 'add_tests' ] );
-		add_filter( 'debug_information', [ $this, 'add_debug_information' ] );
-		add_filter( 'site_status_test_result', [ $this, 'modify_test_result' ] );
-		add_filter( 'site_status_test_php_modules', [ $this, 'add_extensions' ] );
-		add_action( 'admin_print_styles-site-health.php', [ $this, 'add_styles' ] );
+
+		if ( is_admin() ) {
+			add_filter( 'site_status_tests', [ $this, 'add_tests' ] );
+			add_filter( 'debug_information', [ $this, 'add_debug_information' ] );
+			add_filter( 'site_status_test_result', [ $this, 'modify_test_result' ] );
+			add_filter( 'site_status_test_php_modules', [ $this, 'add_extensions' ] );
+			add_action( 'admin_print_styles-site-health.php', [ $this, 'add_styles' ] );
+		}
+
+		add_action( 'template_redirect', [ $this, 'render_random_string_page' ] );
```

Comment:
This could be limited to only users who have the ability to view Site Health checks:

```suggestion
		if ( current_user_can( 'view_site_health_checks' ) ) {
			add_action( 'template_redirect', [ $this, 'render_random_string_page' ] );
		}
```

Reply:
@pierlon We should not add that check. Since idea is to check if page cache is enabled for end-user or not. And for that reason, we make a plain request (without any cookies). If we add that check then when we make a request that won't execute.

Reply:
The only time we would need this request to execute is to check if page caching works, no?

> If we add that check then when we make a request that won't execute.

It won't execute for users who can't view the Site Health checks, which I think would be valid in this case. One problem I did not recognize at the time was that if page caching was enabled it would be possible for a non-authenticated user to request that URL and cache the home page instead of the random number. That could be resolved by using a random query param value (like `home_url( '?amp_page_cache=' . wp_rand( 1000, 10000 ) )` instead of `home_url( '?amp_page_cache=1' )`.

Reply:
@pierlon I think this does need to be added for unauthenticated users. The site health check is run by an authenticated user, but the requests are made in an unauthenticated context. So the `template_redirect` needs to be added for unauthenticated users.

----

@dhaval-parekh After thinking about this some more, I think the use of a query parameter will be problematic a some caching plugins prevent all caching altogether when query parameters are present. 

So here's my suggestion: 

1. Use a HTTP request header to request the random string, and send that random string back via an HTTP response header rather than dying.
2. Even better, perhaps the entire random variable in the response was a bad idea from me to begin with. To detect page caching, should we not rather be checking the `Expires` response header that in the future or that `Cache-Control` that has a non-zero `max-age`? Alongside `Expires`/`Cache-Control` we can also check if the response includes an `ETag` header and/or a `Last-Modified` header, which are also signals that the response is cacheable.

If doing the latter, we don't need to do multiple requests. We can just do one request to the homepage, without any cookies and without any custom request headers or query parameters. If that one request returns with `Expires`/`Cache-Control`/`Last-Modified`/`ETag` that indicates caching is page enabled, then that all we need.

This presumes that page caching always will include HTTP response headers that tell the client they can cache the pages as well, which I think is a fair assumption. Caching plugins are “behooved” to send such response headers as it allows other caching layers to cache and serve back responses without requiring hits to origin.

Nevertheless, if it turns out that caching plugins don't send back response headers for client-cacheability, then there will still be the need to do multiple requests as in option 1 above, where the client can send a `X-AMP-Request-Random-Number: 1` request header, and then the response can include an `X-AMP-Random-Number: ...` response header. By using HTTP headers, we can avoid false negatives for detection while also avoiding accidentally polluting the cache by using harmless HTTP headers.

Reply:
I think both of these approaches are needed actually. The Site Health test can have two degrees of passing:

1. Server-caching: Repeated unauthenticated requests to the server return the same response (by checking the random response header).
2. Client-caching: Server sends back `Cache-Control`/`Expires`/`Last-Modified`/`ETag` headers.

_Both_ of these should be checked because both will result in a better page experience. For 1, this reduces the TTFB for users requesting a page even if they have not visited before. For 2, this allows browsers to serve from disk cache a page that they previously visited, either without hitting the server at all (if within the `max-age`) or by making a conditional request (sending `If-Modified-Since` or `If-None-Match` requests headers).

The first check should be _critical_ in terms of priority. Otherwise, if that passes, the second one would be _recommended_.

Reply:
Note: If repeated requests with an `X-AMP-Request-Random-Number` request header result in responses _without_ a `X-AMP-Random-Number` response header, it could be that there is a caching layer that is stripping that out. In such case, that could indicate that caching is working and the client-caching check should be used to confirm that server-caching is indeed working.